Edited Aevol/doc/aevol_propagate.1
\"test with man -l <file>
.TH AEVOL "1" "July 2014" "aevol 4.4" "User Manual"
.SH NAME
aevol \- an in silico experimental evolution platform
.SH SYNOPSIS
.B aevol_propagate \-h
or
.B \-\-help
.br
.B aevol_propagate \-V
or
.B \-\-version
.br
.B aevol_propagate \fR[\fB\-v\fR]  \fR[\fB\-g\fI GENER\fR]  \fR[\fB\-i\fI INDIR\fR] \fR[\fB\-o\fI OUTDIR\fR] \fR[\fB\-S\fI GENERALSEED\fR]
.br
.B aevol_propagate \fR[\fB\-v\fR]  \fR[\fB\-g\fI GENER\fR]  \fR[\fB\-i\fI INDIR\fR] \fR[\fB\-o\fI OUTDIR\fR] \fR[\fB\-s\fI SELSEED\fR]   \fR[\fB\-m\fI MUTSEED\fR]   \fR[\fB\-t\fI STOCHSEED\fR]   \fR[\fB\-e\fI ENVVARSEED\fR]   \fR[\fB\-n\fI ENVNOISESEED\fR]
.SH DESCRIPTION
.B Aevol
is a simulation platform that allows one to let populations of digital organisms evolve in different conditions and study experimentally the mechanisms responsible for the structuration of the genome and the transcriptome.
.TP
.B aevol_propagate
creates a fresh copy of the experiment as it was at the given generation. The generation number of the copy will be reset to 0. If you use aevol_propagate repeatedly to initialize several simulations, you should specify a different seed for each simulation, otherwise all simulations will yield exactly the same results. You can use the option -S to do so. In this case, the random drawings will be different for all random processes enabled in your simulations (mutations, stochastic gene expression, selection, migration, environmental variation, environmental noise). Alternatively, to change the random drawings for specific random processes only, do not use -S but the options -m, -s, -t, -e, -n (see below).
.SH OPTIONS
.TP
.B \-h, \-\-help
print help, then exit
.TP
.B \-V, \-\-version
print version number, then exit
.TP
.B \-v, \-\-verbose
be verbose
.HP
.B \-g, \-\-gener
.I GENER
.br
specify generation number
(default: that contained in file last_gener.txt, if any)
.HP
.B \-i, \-\-in
.I INDIR
.br
specify input directory (default ".")
.HP
.B \-o, \-\-out
.I OUTDIR
.br
specify output directory (default "./output")
.HP
.B \-S, \-\-general\-seed
.I GENERALSEED
.br
specify an integer to be used as a seed for random numbers. If you use aevol_propagate repeatedly to initialize several simulations, you should specify a different	seed for each simulation, otherwise all simulations will yield exactly the same results. If you specify this general seed, random drawings will be different for all random processes enabled in your simulations (mutations, stochastic gene expression, selection, migration, environmental variation, environmental noise). To change the random drawings for a specific random process only, do not use -S but the options below.
.HP
.B \-s, \-\-sel\-seed
.I SELSEED
.br
specify an integer as a seed for random numbers needed for selection and migration (if spatial structure is enabled).
.HP
.B \-m, \-\-mut\-seed
.I MUTSEED
.br
specify an integer as a seed for random numbers needed for mutations.
.HP
.B \-t, \-\-stoch\-seed
.I STOCHSEED
.br
specify an integer as a seed for random numbers needed for stochastic gene expression.
.HP
.B \-e, \-\-env\-var\-seed
.I ENVVARSEED
.br
specify an integer as a seed for random numbers needed for environmental variation.
.HP
.B \-n, \-\-env\-noise\-seed
.I ENVNOISESEED
.br
specify an integer as a seed for random numbers needed for environmental noise.
.SH "SEE ALSO"
.B aevol_create, aevol_run, aevol_modify


Edited Aevol/ChangeLog
New in 8.0:
 - MPI Support:
    - Fully reproduceable MPI execution with sequential and memory shared parallel version (OpenMP)
    - Can be mixed up with OpenMP version: MPI+OpenMP
    - No (Explicit) Synchronization (i.e. no MPI Barrier). Should be very scalable.
    - RAevol (dynamic and static environment) and Aevol compatible
    - Lineage/Tree are fully supported
    
 - RAevol should be fixed
 - A lot of vectorization works on motif searchs function
 - Validated discrete fuzzy (against the continious fuzzy). WARNING : never compare results of discrete and continious fuzzy
 - A lot of memory leaks fixes
 - A lot of bug fixes
 - New differential search algorithm that does the diff for all metadatas (not only promoters all the metadatas Promoters/RNAs/Genes)

New in 4.4:

* aevol_create now gives a name to the created strain. It can be defined using
  the STRAIN_NAME keyword in the param file. If this keyword is not present, a
  default one will be generated.

* New option -o or --out in aevol_create
  Allows one to create the experiment in any location.

* aevol_propagate can now reset the PRNGs (random generators) with the new
  general option -S (affecting all random processes) or the new more specific
  options -m, -s, -t, -e and -n (each affects a single random process, see man
  page).

* The environmental target can be defined by a list of custom points using the
  ENV_ADD_POINT keyword in the parameter file. This is an alternative to the
  usual way, where it is defined as a sum of gaussians. Note, however, that
  environmental variation is possible only for the "gaussian" way.

* The aevol_misc_fixed_mutations program now outputs the number of genes
  affected by each mutation. More precisely, three new columns are produced:
    - the number of coding RNAs possibly disrupted by a switch, a small indel
      or a rearrangement breakpoint,
    - the number of coding RNAs entirely included in a rearranged segment,
    - in the case of a transfer by replacement, the number of genes that were
      entirely included in the replaced segment.
  This program does not work yet if plasmids are allowed.

* The programs aevol_misc_lineage, aevol_misc_ancstats,
  aevol_misc_fixed_mutations, aevol_misc_gene_families now work when the
  environment was changed at some point by aevol_modify.

* The fitness proportionate selection scheme now works with spatial structure

* In aevol_modify, it is now possible to replace a population by clones of the
  best individual. To do so, add the line CLONE_BEST_NO_TREE or
  CLONE_BEST_AND_CHANGE_TREE in the parameter file you supply to aevol_modify.
  Use CLONE_BEST_AND_CHANGE_TREE if your simulation campaign uses the
  genealogical trees, that is to say if you set RECORD_TREE to true. Otherwise,
  use CLONE_BEST_NO_TREE.

* Manual pages were added for aevol_misc_lineage, aevol_misc_ancstats,
  aevol_misc_fixed_mutations, aevol_misc_gene_families, aevol_misc_create_eps,
  aevol_misc_view_generation.

* New option (-c and -p) in aevol_create. Allows one to load a chromosome (and
  possibly plasmid) from a text file. If -p is used, ALLOW_PLASMIDS must be set
  to true and -c must also be used. If -c is used and ALLOW_PLASMIDS is set to
  true, -p must also be used.

* Add a script (src/misc/movies.py) to produce a movie from an aevol simulation
  with dumps activated.
  run ./src/misc/movies.py -h for help.
  Needs ffmpeg in path.

* The aevol_misc_gene_families program issues the detailed history of each gene
  family on the lineage of a given individual (providing its lineage file).
  A gene family is defined here as a set of coding sequences that arised by
  duplications of a single original gene. The original gene, called the root of
  the family, can either be one of the genes in the initial ancestor, or a new
  gene created from scratch (for example by a local mutation that transformed a
  non-coding RNA into a coding RNA). The history of gene duplications, gene
  losses and gene mutations in each gene family is represented by a binary tree.
  The program starts by loading the initial genome at the beginning of the
  lineage and by tagging each gene in this initial genome. Each of these initial
  genes is marked as the root of a gene family. Then, each mutation recorded in
  the lineage file is replayed and the fate of all tagged genes is followed and
  recorded in their respective families. When a gene is duplicated, the
  corresponding node in one of the gene trees becomes an internal node, and two
  children nodes are added to it, representing the two gene copies. When a gene
  sequence is modified, the mutation is recorded in its corresponding node in
  one of the gene trees. When a gene is lost, the corresponding node in one of
  the gene trees is labelled as lost. When a new gene appears from scratch, i.e.
  not by gene duplication, it becomes the root of a new gene tree. Environmental
  variations are also replayed exactly as they occured during the main run.
  When all mutations have been replayed, several output files are written in a
  directory called gene_trees. Two general text files are produced. The file
  called gene_tree_statistics.txt contains general data on each gene family,
  like its creation date, its extinction date, or how many nodes it contained.
  The file called nodeattr_tabular.txt contains information about each node of
  each gene tree, like when it was duplicated or lost or how many mutations
  occurred on its branch. In addition, for each gene tree, two text files are
  generated: a file called genetree******-topology.tre contains the topology of
  the gene tree in the Newick format, and a file called
  genetree******-nodeattr.txt that contains the detailed list of events that
  happened to each node in the tree file, before it was either duplicated or
  lost.

  Usage: ae_misc_gene_families [-c | -n] [-t tolerance] -f lineage_file


Changes in 4.4:


Edited Aevol/src/aevol_propagate.cpp
// ****************************************************************************
//
//          Aevol - An in silico experimental evolution platform
//
// ****************************************************************************
//
// Copyright: See the AUTHORS file provided with the package or <www.aevol.fr>
// Web: http://www.aevol.fr/
// E-mail: See <http://www.aevol.fr/contact/>
// Original Authors : Guillaume Beslon, Carole Knibbe, David Parsons
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 2 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//
// ****************************************************************************


const char* DEFAULT_PARAM_FILE_NAME = "param.in";

// ============================================================================
//                                   Includes
// ============================================================================
#include <cstdlib>
#include <cstdio>
#include <cstring>
#include <err.h>
#include <errno.h>
#include <getopt.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <sys/stat.h>

#include <limits>

// TODO(dpa) will be merged into C++17, remove dependency when possible
#include <boost/filesystem.hpp>


// =================================================================
//                            Project Files
// =================================================================
#if __cplusplus == 201103L
  #include "make_unique.h"
#endif

#ifdef __X11
  #include "ExpManager_X11.h"
#else
  #include "ExpManager.h"
#endif

#include "ParamLoader.h"

using namespace aevol;

// Helper functions
void print_help(char* prog_path);
void interpret_cmd_line_options(int argc, char* argv[]);

// Command-line option variables
static int64_t timestep = -1;
static char* input_dir = nullptr;
static char* output_dir = nullptr;
static bool keep_prng_states = false;
static bool verbose = false;

static int32_t selseed        = -1;

int main(int argc, char* argv[]) {
  interpret_cmd_line_options(argc, argv);

  // If output directory doesn't exist, create it
  struct stat stat_buf;
  if ((stat(output_dir, &stat_buf) == -1) && (errno == ENOENT)) {
    if (not boost::filesystem::create_directories(output_dir)) {
      err(EXIT_FAILURE, output_dir, errno);
    }
  }



  // =================================================================
  //                    Load the model experiment
  // =================================================================
#ifndef __NO_X
  ExpManager *exp_manager = new ExpManager_X11();
#else
  ExpManager* exp_manager = new ExpManager();
#endif

  exp_manager->load(input_dir, timestep, verbose, true);

  if (not keep_prng_states) {

      auto max = std::numeric_limits<int32_t>::max();

#if __cplusplus == 201103L
    auto prng = make_unique<JumpingMT>(time(nullptr));

    exp_manager->sel()->set_prng(
        make_unique<JumpingMT>(prng->random(max)));
    exp_manager->world()->set_prng(
        make_unique<JumpingMT>(prng->random(max)));
#else
    auto prng = std::make_unique<JumpingMT>(time(nullptr));

    for (int16_t x = 0; x < exp_manager->world()->width(); x++) {
      for (int16_t y = 0; y < exp_manager->world()->height(); y++) {
        int32_t seed = prng->random(1000000);
#if __cplusplus == 201103L
        exp_manager->world()->grid(x,y)->set_reprod_prng(make_unique<JumpingMT>(seed));
          exp_manager->world()->grid(x,y)->set_reprod_prng_simd(make_unique<JumpingMT>(seed));
          exp_manager->world()->grid(x,y)->set_mut_prng(std::make_shared<JumpingMT>(seed));
          exp_manager->world()->grid(x,y)->set_stoch_prng(std::make_shared<JumpingMT>(seed));
#else
        exp_manager->world()->grid(x, y)->set_reprod_prng(std::make_unique<JumpingMT>(seed));
        exp_manager->world()->grid(x, y)->set_reprod_prng_simd(std::make_unique<JumpingMT>(seed));
        exp_manager->world()->grid(x, y)->set_mut_prng(std::make_shared<JumpingMT>(seed));
        exp_manager->world()->grid(x, y)->set_stoch_prng(std::make_shared<JumpingMT>(seed));
#endif
      }
    }
#endif


    exp_manager->world()->set_phen_target_prngs(
            std::make_shared<JumpingMT>(prng->random(max)),
            std::make_shared<JumpingMT>(prng->random(max)));

      exp_manager->save_copy(output_dir, timestep);

  } else {

    if (selseed != -1) {
#if __cplusplus == 201103L
      auto prng = make_unique<JumpingMT>(selseed);
#else
      auto prng = std::make_unique<JumpingMT>(selseed);
#endif

      for (int16_t x = 0; x < exp_manager->world()->width(); x++) {
        for (int16_t y = 0; y < exp_manager->world()->height(); y++) {
          int32_t seed = prng->random(1000000);
#if __cplusplus == 201103L
          exp_manager->world()->grid(x,y)->set_reprod_prng(make_unique<JumpingMT>(seed));
          exp_manager->world()->grid(x,y)->set_reprod_prng_simd(make_unique<JumpingMT>(seed));
#else
          exp_manager->world()->grid(x, y)->set_reprod_prng(std::make_unique<JumpingMT>(seed));
          exp_manager->world()->grid(x, y)->set_reprod_prng_simd(std::make_unique<JumpingMT>(seed));
#endif
        }
      }

    }


      exp_manager->save_copy(output_dir, timestep);
  }
}







/*!
  \brief print help and exist

*/
    void print_help(char *prog_path) {
      // Get the program file-name in prog_name (strip prog_path of the path)
      char *prog_name; // No new, it will point to somewhere inside prog_path
      if ((prog_name = strrchr(prog_path, '/'))) {
        prog_name++;
      } else {
        prog_name = prog_path;
      }

      printf("******************************************************************************\n");
      printf("*                                                                            *\n");
      printf("*                        aevol - Artificial Evolution                        *\n");
      printf("*                                                                            *\n");
      printf("* Aevol is a simulation platform that allows one to let populations of       *\n");
      printf("* digital organisms evolve in different conditions and study experimentally  *\n");
      printf("* the mechanisms responsible for the structuration of the genome and the     *\n");
      printf("* transcriptome.                                                             *\n");
      printf("*                                                                            *\n");
      printf("******************************************************************************\n");
      printf("\n");
      printf("%s:\n", prog_name);
      printf("\tCreate a fresh copy of the experiment as it was at the given timestep.\n");
      printf("\tThe timestep number of the copy will be reset to 0.\n");
      printf("\n");
      printf("Usage : %s -h or --help\n", prog_name);
      printf("   or : %s -V or --version\n", prog_name);
      printf("   or : %s [-t TIMESTEP] [-K] [-o OUTDIR] [-v]\n",
             prog_name);
      printf("\nOptions\n");
      printf("  -h, --help\n\tprint this help, then exit\n");
      printf("  -V, --version\n\tprint version number, then exit\n");
      printf("  -t, --timestep TIMESTEP\n");
      printf("\tspecify timestep to propagate\n");
      printf("  -K, --keep-prng-st\n\tdo not alter prng states\n");
//  printf("  -i, --in INDIR\n"
//             "\tspecify input directory (default \".\")\n");
      printf("  -o, --out OUTDIR\n"
             "\tspecify output directory (default \"./output\")\n");
      printf("  -v, --verbose\n\tbe verbose\n");
    }

    void interpret_cmd_line_options(int argc, char *argv[]) {
      // Define allowed options
      const char *options_list = "hVt:o:v";
      static struct option long_options_list[] = {
              {"help",         no_argument,       nullptr, 'h'},
              {"version",      no_argument,       nullptr, 'V'},
              {"timestep",     required_argument, nullptr, 't'},
              {"keep-prng-st", no_argument,       nullptr, 'K'},
//      {"in",             required_argument, nullptr, 'i'},
              {"out",          required_argument, nullptr, 'o'},
              {"verbose",      no_argument,       nullptr, 'v'},
              {0, 0,                              0,       0}
      };

      // Get actual values of the CLI options
      int option;
      while ((option = getopt_long(argc, argv, options_list, long_options_list,
                                   nullptr)) != -1) {
        switch (option) {
          case 'h' : {
            print_help(argv[0]);
            exit(EXIT_SUCCESS);
          }
          case 'V' : {
            Utils::PrintAevolVersion();
            exit(EXIT_SUCCESS);
          }
          case 't' : {
            timestep = atoi(optarg);
            break;
          }
          case 'K' : {
            keep_prng_states = true;
            break;
          }
//      case 'i' : {
//        input_dir = new char[strlen(optarg) + 1];
//        strcpy(input_dir, optarg);
//        break;
//      }
          case 'o' : {
            output_dir = new char[strlen(optarg) + 1];
            strcpy(output_dir, optarg);
            break;
          }
          case 'v' : {
            verbose = true;
            break;
          }
          default : {
            // An error message is printed in getopt_long, we just need to exit
            exit(EXIT_FAILURE);
          }
        }
      }

      // If input directory wasn't provided, use default
      if (input_dir == nullptr) {
        input_dir = new char[255];
        sprintf(input_dir, "%s", ".");
      }
      // If output directory wasn't provided, use default
      if (output_dir == nullptr) {
        output_dir = new char[255];
        sprintf(output_dir, "%s", "output");
      }

      // If timestep wasn't provided, use default
      if (timestep == -1) {
        timestep = OutputManager::last_gener();
      }
    }

